

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>VirtualMicrobes.post_analysis.network_properties &mdash; Virtual Microbes Evolutionary Simulator 0.1.3 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../../genindex.html"/>
        <link rel="search" title="Search" href="../../../search.html"/>
    <link rel="top" title="Virtual Microbes Evolutionary Simulator 0.1.3 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> Virtual Microbes Evolutionary Simulator
          

          
          </a>

          
            
            
              <div class="version">
                0.1.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../reference/index.html">Reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">Virtual Microbes Evolutionary Simulator</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>VirtualMicrobes.post_analysis.network_properties</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for VirtualMicrobes.post_analysis.network_properties</h1><div class="highlight"><pre>
<span></span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">Created on Feb 9, 2016</span>

<span class="sd">@author: thocu</span>
<span class="sd">&#39;&#39;&#39;</span>
<span class="kn">from</span> <span class="nn">VirtualMicrobes.virtual_cell</span> <span class="k">import</span> <span class="n">PhyloUnit</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">OrderedDict</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<div class="viewcode-block" id="PhyloGeneticAnalysis"><a class="viewcode-back" href="../../../reference/VirtualMicrobes.post_analysis.html#VirtualMicrobes.post_analysis.network_properties.PhyloGeneticAnalysis">[docs]</a><span class="k">class</span> <span class="nc">PhyloGeneticAnalysis</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Analyze biological networks</span>
<span class="sd">    &#39;&#39;&#39;</span>

    
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">        </span>
<span class="sd">        &#39;&#39;&#39;</span>
        <span class="k">pass</span></div>
        

<div class="viewcode-block" id="find_homologs"><a class="viewcode-back" href="../../../reference/VirtualMicrobes.post_analysis.html#VirtualMicrobes.post_analysis.network_properties.find_homologs">[docs]</a><span class="k">def</span> <span class="nf">find_homologs</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For a gene, find all its homologs in a given genome.</span>
<span class="sd">    </span>
<span class="sd">    This is a naive approach that uses a combination of the gene&#39;s type and its</span>
<span class="sd">    :class:`VirtualMicrobes.virtual_cell.Identifier.Identifier` attribute to</span>
<span class="sd">    detect common descent.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene : :class:`VirtualMicrobes.virtual_cell.GenomicElement.GenomicElement`</span>
<span class="sd">        Reference gene for homology search.</span>
<span class="sd">    genome : :class:`VirtualMicrobes.virtual_cell.Genome.Genome`</span>
<span class="sd">        Genome in which to search for homologs.</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    The set of homologs of `gene` in the `genome`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">homologous</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="o">==</span> <span class="nb">type</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="ow">and</span> <span class="n">gene</span><span class="o">.</span><span class="n">id</span><span class="o">.</span><span class="n">is_copy</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">id</span><span class="p">)):</span>
            <span class="n">homologous</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">homologous</span></div>
        
<div class="viewcode-block" id="find_homolog_distances"><a class="viewcode-back" href="../../../reference/VirtualMicrobes.post_analysis.html#VirtualMicrobes.post_analysis.network_properties.find_homolog_distances">[docs]</a><span class="k">def</span> <span class="nf">find_homolog_distances</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">closest_homolog</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find homologs and their distance for a gene in a target genome.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene : :class:`VirtualMicrobes.virtual_cell.GenomicElement.GenomicElement`</span>
<span class="sd">        Reference gene for homology search.</span>
<span class="sd">    genome : :class:`VirtualMicrobes.virtual_cell.Genome.Genome`</span>
<span class="sd">        Genome in which to search for homologs.</span>
<span class="sd">    closest_homolog : bool</span>
<span class="sd">        Flag to filter found homologs to those that have the shortest</span>
<span class="sd">        phylogenetic distance to the `gene`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">age</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">time_birth</span>
    <span class="c1">#print &#39;gene age&#39;, age</span>
    <span class="n">homolog_distance_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">find_homologs</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">genome</span><span class="p">):</span>
        <span class="n">dist</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">h_age</span> <span class="o">=</span> <span class="n">h</span><span class="o">.</span><span class="n">time_birth</span>
        <span class="k">if</span> <span class="n">h</span> <span class="ow">is</span> <span class="n">gene</span><span class="p">:</span>
            <span class="n">dist</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">homolog_distance_dict</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
        <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">parent_of</span><span class="p">(</span><span class="n">h</span><span class="p">):</span>
            <span class="n">dist</span> <span class="o">=</span> <span class="n">age</span> <span class="o">-</span> <span class="n">h_age</span>
            <span class="n">homolog_distance_dict</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
            <span class="c1">#print &#39;found parent&#39;, h, &#39;at distance&#39;, dist</span>
        <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">child_of</span><span class="p">(</span><span class="n">h</span><span class="p">):</span>
            <span class="n">dist</span> <span class="o">=</span> <span class="n">h_age</span> <span class="o">-</span> <span class="n">age</span>
            <span class="n">homolog_distance_dict</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>
            <span class="c1">#print &#39;found child&#39;, h, &#39;at distance&#39;, dist</span>
        <span class="k">else</span><span class="p">:</span> 
            <span class="n">common_ancestors</span> <span class="o">=</span> <span class="n">gene</span><span class="o">.</span><span class="n">common_ancestors</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
            <span class="n">ca_dists</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">ca</span> <span class="ow">in</span> <span class="n">common_ancestors</span><span class="p">:</span>
                <span class="n">ca_dists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">age</span> <span class="o">+</span> <span class="n">h_age</span> <span class="o">-</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">ca</span><span class="o">.</span><span class="n">time_birth</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ca_dists</span><span class="p">):</span>
                <span class="n">homolog_distance_dict</span><span class="p">[</span><span class="n">h</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">ca_dists</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">closest_homolog</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">homolog_distance_dict</span><span class="p">):</span>
        <span class="n">min_dist</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">homolog_distance_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
        <span class="n">homolog_distance_dict</span> <span class="o">=</span> <span class="p">{</span> <span class="n">k</span><span class="p">:</span><span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">homolog_distance_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> 
                                 <span class="k">if</span> <span class="n">v</span> <span class="o">==</span> <span class="n">min_dist</span> <span class="p">}</span>
    <span class="k">return</span> <span class="n">homolog_distance_dict</span></div>
     
<div class="viewcode-block" id="calculate_overlap"><a class="viewcode-back" href="../../../reference/VirtualMicrobes.post_analysis.html#VirtualMicrobes.post_analysis.network_properties.calculate_overlap">[docs]</a><span class="k">def</span> <span class="nf">calculate_overlap</span><span class="p">(</span><span class="n">tf_connections</span><span class="p">,</span> <span class="n">connections_of_homologous_tfs</span><span class="p">,</span> 
                      <span class="n">closest_bound_homologs_dict</span><span class="p">):</span>   
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the overlap in bound genes between tf homologs.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    tf_connections : list of :class:`VirtualMicrobes.virtual_cell.Gene.Gene`\s</span>
<span class="sd">        Downstream connections of the reference TF</span>
<span class="sd">    connections_of_homologous_tfs : list of sets of :class:`VirtualMicrobes.virtual_cell.Gene.Gene`\s</span>
<span class="sd">        List of sets of downstream genes for each homolog of the reference TF  </span>
<span class="sd">    closest_bound_homologs_dict : dict of sets of :class:`VirtualMicrobes.virtual_cell.Gene.Gene`\s</span>
<span class="sd">        Mapping of each original downstream gene of the reference TF to sets of</span>
<span class="sd">        homologs of these downstream genes.</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    float,float</span>
<span class="sd">        Tuple of fractions: </span>
<span class="sd">        [0]: Fraction of downstream genes who&#39;s homologs are bound by a homolog of</span>
<span class="sd">        the reference TF.</span>
<span class="sd">        [1]: Fraction of new connections (averaged over tf homologs) per original </span>
<span class="sd">        connection of the reference TF.  </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">connections_still_bound</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">nr_homologous_tfs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">connections_of_homologous_tfs</span><span class="p">)</span>
    <span class="n">all_closest_homolog_of_bound_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">all_connections_of_homologous_tfs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">bound_gene</span> <span class="ow">in</span> <span class="n">tf_connections</span><span class="p">:</span>
        <span class="n">closest_homologs_of_bound_gene</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">closest_bound_homologs_dict</span><span class="p">[</span><span class="n">bound_gene</span><span class="p">])</span>
        <span class="n">all_closest_homolog_of_bound_genes</span> <span class="o">|=</span> <span class="n">closest_homologs_of_bound_gene</span>
        <span class="n">still_bound</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">connections</span> <span class="ow">in</span> <span class="n">connections_of_homologous_tfs</span><span class="p">:</span>
            <span class="n">overlap</span> <span class="o">=</span> <span class="n">connections</span> <span class="o">&amp;</span> <span class="n">closest_homologs_of_bound_gene</span>
            <span class="n">all_connections_of_homologous_tfs</span> <span class="o">|=</span> <span class="n">connections</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span> <span class="c1"># the homolog of the tf binds a homolog of the bound gene</span>
                <span class="n">still_bound</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">still_bound</span><span class="p">:</span>
            <span class="n">connections_still_bound</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="n">all_new</span> <span class="o">=</span> <span class="n">all_connections_of_homologous_tfs</span> <span class="o">-</span> <span class="n">all_closest_homolog_of_bound_genes</span>
    <span class="n">fraction_conserved</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span> 
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tf_connections</span><span class="p">):</span>
        <span class="n">fraction_conserved</span> <span class="o">=</span> <span class="n">connections_still_bound</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tf_connections</span><span class="p">))</span>
    
    <span class="n">fraction_new</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">NaN</span> 
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tf_connections</span><span class="p">)</span> <span class="ow">and</span> <span class="n">nr_homologous_tfs</span><span class="p">:</span>
        <span class="n">avrg_new_connections</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">all_new</span><span class="p">)</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="n">nr_homologous_tfs</span><span class="p">)</span>
        <span class="n">fraction_new</span> <span class="o">=</span> <span class="n">avrg_new_connections</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tf_connections</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">fraction_conserved</span> <span class="p">,</span> <span class="n">fraction_new</span></div>
     
<div class="viewcode-block" id="tf_binding_overlap"><a class="viewcode-back" href="../../../reference/VirtualMicrobes.post_analysis.html#VirtualMicrobes.post_analysis.network_properties.tf_binding_overlap">[docs]</a><span class="k">def</span> <span class="nf">tf_binding_overlap</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">,</span> <span class="n">closest_homolog</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">no_phylogeny</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Measure the overlap in target genes for tf homologs in phylogenetically</span>
<span class="sd">    related individuals.</span>
<span class="sd">    </span>
<span class="sd">    cell1 : :class:`VirtualMicrobes.virtual_cell.Cell.Cell`</span>
<span class="sd">        Reference individual for which to find homologs</span>
<span class="sd">    cell2 : :class:`VirtualMicrobes.virtual_cell.Cell.Cell`</span>
<span class="sd">        Homologs of TFs and downstream targets will be detected in this individual. </span>
<span class="sd">    closest_homolog : bool</span>
<span class="sd">        Flag to filter found homologs to those that have the shortest</span>
<span class="sd">        phylogenetic distance to the `gene`.</span>
<span class="sd">    verbose : bool</span>
<span class="sd">        Print messages about homologs found.</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    dict</span>
<span class="sd">        Mapping from :class:`VirtualMicrobes.virtual_cell.Gene.TranscriptionFactor` to (maximum)</span>
<span class="sd">        binding overlap score.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">overlap_dict</span><span class="p">,</span> <span class="n">new_dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(),</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">cell1_tf_connections</span> <span class="o">=</span> <span class="n">cell1</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">tf_connections_dict</span><span class="p">()</span>
    <span class="n">cell2_tf_connections</span> <span class="o">=</span> <span class="n">cell2</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">tf_connections_dict</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span> <span class="s1">&#39;cell1&#39;</span><span class="p">,</span> <span class="n">cell1</span><span class="o">.</span><span class="n">time_birth</span><span class="p">,</span> <span class="s1">&#39;-&gt; cell2&#39;</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">time_birth</span>
    <span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">cell1</span><span class="o">.</span><span class="n">genome</span><span class="o">.</span><span class="n">tfs</span><span class="p">):</span>
        <span class="n">bound_genes</span> <span class="o">=</span> <span class="n">cell1_tf_connections</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span><span class="n">PhyloUnit</span><span class="o">.</span><span class="n">PhyloUnit</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">no_phylogeny</span><span class="p">:</span>
            <span class="c1">#print &#39;using detailed phylogenetic distance information&#39;</span>
            <span class="n">closest_tf_homologs</span> <span class="o">=</span> <span class="n">find_homolog_distances</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">closest_homolog</span><span class="p">)</span>
            <span class="n">closest_bound_homolog_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">bg</span> <span class="ow">in</span> <span class="n">bound_genes</span><span class="p">:</span>
                <span class="n">closest_bound_homolog_dict</span><span class="p">[</span><span class="n">bg</span><span class="p">]</span> <span class="o">=</span> <span class="n">find_homolog_distances</span><span class="p">(</span><span class="n">bg</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">genome</span><span class="p">,</span> <span class="n">closest_homolog</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#print &#39;No phylogeny stored for object. Using simple homology by identifier information.&#39;</span>
            <span class="n">closest_tf_homologs</span> <span class="o">=</span> <span class="n">find_homologs</span><span class="p">(</span><span class="n">tf</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span> 
            <span class="n">closest_bound_homolog_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">bg</span> <span class="ow">in</span> <span class="n">bound_genes</span><span class="p">:</span>
                <span class="n">closest_bound_homolog_dict</span><span class="p">[</span><span class="n">bg</span><span class="p">]</span> <span class="o">=</span> <span class="n">find_homologs</span><span class="p">(</span><span class="n">bg</span><span class="p">,</span> <span class="n">cell2</span><span class="o">.</span><span class="n">genome</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">found&#39;</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">homologs</span><span class="p">)</span> <span class="k">for</span> <span class="n">homologs</span> <span class="ow">in</span> <span class="n">closest_bound_homolog_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="p">]),</span>
            <span class="nb">print</span> <span class="s1">&#39;homologs for&#39;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">bound_genes</span><span class="p">),</span> <span class="s1">&#39;downstream genes bound now&#39;</span>
        
        <span class="n">overlap_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">],</span> <span class="n">new_dict</span><span class="p">[</span><span class="n">tf</span><span class="p">]</span> <span class="o">=</span> <span class="n">calculate_overlap</span><span class="p">(</span><span class="n">bound_genes</span><span class="p">,</span> 
                                             <span class="p">[</span> <span class="n">cell2_tf_connections</span><span class="p">[</span><span class="n">h_tf</span><span class="p">]</span> 
                                              <span class="k">for</span> <span class="n">h_tf</span> <span class="ow">in</span> <span class="n">closest_tf_homologs</span> <span class="p">],</span> 
                                             <span class="n">closest_bound_homolog_dict</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span>
    <span class="k">return</span> <span class="n">overlap_dict</span><span class="p">,</span> <span class="n">new_dict</span></div>
</pre></div>

           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Thomas Cuypers, Bram van Dijk.
      Last updated on Oct 10, 2016.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.1.3',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>